### PR TITLE
[BEAM-23] Refactors to support key-preserving in direct runner

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

R: @tgroh 

I've peeled these commits off since they make some sense on their own. The intent is that just tweaking the logic in `KeyedPValueTrackingVisitor` (in potentially hacky ways :-) we can start to allow key preservation.

### CODE REVIEW

**Comment by kennknowles** [comment]
retest this please

**Comment by kennknowles** [comment]
Failure was uploading to Coveralls after success.

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/KeyedPValueTrackingVisitorTest.java [hunk 1]
```diff
@@ -117,26 +80,17 @@ public void noInputUnkeyedOutput() {
   }
 
   @Test
-  public void keyedInputNotProducesKeyedOutputUnkeyedOutput() {
-    PCollection<Integer> onceKeyed =
-        p.apply(Create.<Integer>of(1, 2, 3))
-            .apply(new PrimitiveKeyer<Integer>())
-            .apply(ParDo.of(new IdentityFn<Integer>()));
+  public void keyedInputWithoutKeyPreserving() {
```
Comment (tgroh) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/KeyedPValueTrackingVisitorTest.java]: `withKeyPreserving`?

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/EvaluationContext.java [hunk 2]
```diff
@@ -99,17 +100,28 @@
 
   private final DirectMetrics metrics;
 
+  private final Set<PValue> keyedPValues;
```
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/EvaluationContext.java]: This might be worth just adding to `DirectGraph` and `DirectGraphVisitor`

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/KeyedPValueTrackingVisitor.java [hunk 3]
```diff
@@ -93,4 +103,9 @@ public void visitValue(PValue value, TransformHierarchy.Node producer) {
         finalized, "can't call getKeyedPValues before a Pipeline has been completely traversed");
     return keyedValues;
   }
+
+  // There are currently no key-preserving transforms; this lays the infrastructure for them
```
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/KeyedPValueTrackingVisitor.java]: I'd move this into the method definition.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluator.java [hunk 4]
```diff
@@ -86,6 +95,10 @@
         evaluationContext, runner, application, aggregatorChanges, outputManager, stepContext);
   }
 
+  private static boolean isKeyPreserving(Serializable fn) {
```
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluator.java]: Unused?

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/KeyedPValueTrackingVisitor.java (hunk 3 above)
Comment (kennknowles) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/KeyedPValueTrackingVisitor.java]: Done

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluator.java (hunk 4 above)
Comment (kennknowles) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluator.java]: Done

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/KeyedPValueTrackingVisitorTest.java (hunk 1 above)
Comment (kennknowles) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/KeyedPValueTrackingVisitorTest.java]: Since there are no key-preserving transforms I can't test it without adding abstractions that we ultimately won't need. Rebased on top of this, I do have a PR that alters stateful `ParDo` to have a key-preserving intermediate function, but I was throwing out this refactor first since it was trivial. I could add all that to this one if you like.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/EvaluationContext.java (hunk 2 above)
Comment (kennknowles) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/EvaluationContext.java]: Seems fine, but scope creep?

**Comment by kennknowles** [comment]
Addressed all comments here, but because of desire to include more motivating tests, superseded by #1581
